stdlib/public/Concurrency/TaskGroup.cpp

```diff
@@ -504,18 +505,37 @@ static void swift_taskGroup_initializeImpl(TaskGroup *group, const Metadata *T)
 }
 
 // =============================================================================
-// ==== add / attachChild ------------------------------------------------------
+// ==== child task management --------------------------------------------------
 
 void TaskGroup::addChildTask(AsyncTask *child) {
   SWIFT_TASK_DEBUG_LOG("attach child task = %p to group = %p", child, this);
 
-  // The counterpart of this (detachChild) is performed by the group itself,
-  // when it offers the completed (child) task's value to a waiting task -
-  // during the implementation of `await group.next()`.
+  // Add the child task to this task group.  The corresponding removal
```

Comment:
I thought we had discussed offline about changing the behaviour so that the child tasks won't remove themselves from the list even after they complete (ie no modification in `offer`) and the parent would do it instead? This would make sure that we don't have contention on both the TaskGroup atomic and the parent's task status record lock?

Reply:
If that's a follow-up change and this is more targeted, that's totally fine. Just wanted to understand the plan.

Reply:
Hmm, I guess that's a tradeoff between memory use of tasks hanging around even though they're completed then, wouldn't it be? Also just thinking out loud here though. Optimization wise it'd be great to revisit and get rid of the lock in the group impl, the status acquire/releases should do enough to guarantee synchronization but I never go to follow through on it :(

Reply:
> Hmm, I guess that's a tradeoff between memory use of tasks hanging around even though they're completed then, wouldn't it be? 

Don't you already have to pay this cost today and keep the children around anyways, since only the first child task which unblocks the parent, can fill the result directly into the parent task? All other child tasks which may have to wait for the parent to consume them, have to hang around anyways. 

> Optimization wise it'd I'd try to revisit and get rid of the lock in the group impl, the status acquire/releases should do enough to guarantee synchronization but I never go to follow through on it :(

I'm not talking about the TaskGroup lock here though? I'm talking about the fact that in this case, we have contention on the TaskGroup atomic as well as on the parent task's task status record lock.

Reply:
@rokhinip The current patch only synchronously removes tasks from the active-tasks list during `offer` if the parent task is currently waiting, so I don't think this should increase contention on the task status atomic unless there's a contending cancelling thread or something.

The memory for a task group child sticks around until the parent awaits that task, so there isn't any difference in memory use here.  Currently that is required because the parent needs the result that's stored in the child's future, which is allocated inline in the task.  That would not be necessary if we separately allocated the future; we'd be able to immediately throw away the rest of the memory.  We could use a more efficient future implementation with task group children anyway because (1) there's only one possible waiter and (2) it's mediated through the task group rather than directly through the task handle.